### PR TITLE
#12558: TTNN implementation of MNIST model

### PR DESCRIPTION
**Pull request by sabira-mcw**
### Ticket
**[Link to GitHub Issue](https://github.com/tenstorrent/tt-metal/issues/12558)**

### Problem description
TTNN implementation for MNIST model.

### What's changed
ttnn support is enabled for mnist model along with demo, e2e and device test perf.

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [x] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/scripts/run_performance.sh [hunk 1]
```diff
@@ -27,6 +27,8 @@ run_perf_models_other() {
 
     env pytest -n auto models/demos/metal_BERT_large_11/tests -m $test_marker
 
+    env pytest -n auto models/experimental/functional_mnist/tests/test_perf_mnist.py -m $test_marke
```
Comment (tt-rkim) [inline tests/scripts/run_performance.sh:30]: Typo

File: models/experimental/functional_mnist/demo/demo.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+from torchvision import transforms, datasets
+from loguru import logger
+import ttnn
+from torch.utils.data import DataLoader
+from models.experimental.mnist.reference.mnist import MnistModel
+from models.experimental.functional_mnist.tt import tt_functional_mnist
+from ttnn.model_preprocessing import preprocess_model_parameters
+
+
+def run_demo_dataset(device, batch_size, iterations, model_location_generator):
```
Comment (tt-rkim) [inline models/experimental/functional_mnist/demo/demo.py:17]: Please post a passing run

File: tests/ttnn/integration_tests/mnist/test_mnist.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import ttnn
+import pytest
+from tests.ttnn.utils_for_testing import assert_with_pcc
+from ttnn.model_preprocessing import preprocess_model_parameters
+from models.experimental.mnist.reference.mnist import MnistModel
+from models.experimental.functional_mnist.tt import tt_functional_mnist
+from torch.utils.data import DataLoader
+from torchvision import transforms, datasets
+
+
```
Comment (tt-rkim) [inline tests/ttnn/integration_tests/mnist/test_mnist.py:15]: Please fast dispatch nightly (which is not very stable) to ensure that your test runs and passes on all 3 card types

File: tests/scripts/run_performance.sh (hunk 1 above)
Comment (sabira-mcw) [reply tests/scripts/run_performance.sh:30]: I will fix it.

**Review by saichandax — APPROVED** [review]

**Review by tt-rkim — APPROVED** [review]

File: models/demos/mnist/demo/demo.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        x, labels = next(iter(dataloader))
+        dataset_predictions = []
+        ttnn_predictions = []
+        dataset_ttnn_correct = 0
+        x = ttnn.from_torch(x, dtype=ttnn.bfloat16, device=device)
+        tt_output = tt_mnist.mnist(device, batch_size, x, parameters)
+        tt_output = ttnn.to_torch(tt_output)
+        predicted_probabilities = torch.nn.functional.softmax(tt_output, dim=1)
+        _, predicted_label = torch.max(predicted_probabilities, 1)
+        tt_output = tt_output
```
Comment (boris-drazic) [inline models/demos/mnist/demo/demo.py:44]: Is this line needed?

File: models/demos/mnist/demo/demo.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            logger.info(f"Iter: {iters} Sample {i}:")
+            logger.info(f"Expected Label: {dataset_predictions[i]}")
+            logger.info(f"Predicted Label: {ttnn_predictions[i]}")
+
+            if dataset_predictions[i] == ttnn_predictions[i]:
+                dataset_ttnn_correct += 1
+                correct += 1
+        dataset_ttnn_accuracy = dataset_ttnn_correct / (batch_size)
+        logger.info(
+            f"ImageNet Inference Accuracy for iter {iters} of {batch_size} input samples : {dataset_ttnn_accuracy}"
```
Comment (boris-drazic) [inline models/demos/mnist/demo/demo.py:57]: I think data set is MNIST not ImageNet

File: models/demos/mnist/demo/demo.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        )
+
+    accuracy = correct / (batch_size * iterations)
+    logger.info(f"ImageNet Inference Accuracy for {batch_size}x{iterations} Samples : {accuracy}")
```
Comment (boris-drazic) [inline models/demos/mnist/demo/demo.py:61]: I think data set is MNIST not ImageNet

File: models/demos/mnist/demo/demo.py (hunk 4 above)
Comment (sabira-mcw) [reply models/demos/mnist/demo/demo.py]: Thank you for pointing that out! I’ve removed the redundant line .

File: models/demos/mnist/demo/demo.py (hunk 6 above)
Comment (sabira-mcw) [reply models/demos/mnist/demo/demo.py]: I’ve corrected the dataset reference to MNIST as you pointed out. Thank you!

**Review by boris-drazic — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

File: models/demos/mnist/tests/test_perf_mnist.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        model_name="MNIST",
+        batch_size=batch_size,
+        inference_and_compile_time=inference_and_compile_time,
+        inference_time=average_inference_time,
+        expected_compile_time=expected_compile_time,
+        expected_inference_time=expected_inference_time,
+        comments="",
+        inference_time_cpu=0.0,
+    )
+
```
Comment (uaydonat) [inline models/demos/mnist/tests/test_perf_mnist.py:92]: test should fail if perf is lower than expected

Comment (sabira-mcw) [reply models/demos/mnist/tests/test_perf_mnist.py:92]: Updated the file with the assert statement.

**Comment by tt-rkim** [comment]
Please post CI links.

**Comment by sabira-mcw** [comment]
CI links:

(Single-card) Demo tests : [n150](https://github.com/tenstorrent/tt-metal/actions/runs/11934249487/job/33263128875#step:9:5038) [n300](https://github.com/tenstorrent/tt-metal/actions/runs/11934249487/job/33263129176), [n300 performance](https://github.com/tenstorrent/tt-metal/actions/runs/11934249487/job/33299923335)(In progress)
All post-commit tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/11934258517) (All Passed)
(Single-card) Model perf tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/11934260943) (All passed)
(Single-card) Device perf regressions : [e150](https://github.com/tenstorrent/tt-metal/actions/runs/11934255399/job/33299918199#step:9:769) , [n300](https://github.com/tenstorrent/tt-metal/actions/runs/11934255399/job/33299918330) (Mnist tests are passing)
Nightly fast dispatch tests : [e150](https://github.com/tenstorrent/tt-metal/actions/runs/11934252391/job/33299804410), [n150](https://github.com/tenstorrent/tt-metal/actions/runs/11934252391/job/33299804714), [n300](https://github.com/tenstorrent/tt-metal/actions/runs/11934252391/job/33299805009) (Mnist tests are passing)

**Comment by tt-rkim** [comment]
* Please fix the threshold for GS: https://github.com/tenstorrent/tt-metal/actions/runs/11934255399/job/33299918199#step:9:753
* Please let us know when demo tests pass

File: models/demos/mnist/tests/test_perf_mnist.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        device=device,
+    )
+    transform = transforms.Compose([transforms.ToTensor()])
+    test_dataset = datasets.MNIST(root="./data", train=False, transform=transform, download=True)
+    dataloader = DataLoader(test_dataset, batch_size=batch_size)
+    x, labels = next(iter(dataloader))
+
+    test_input = ttnn.from_torch(x, dtype=ttnn.bfloat16, device=device)
+    durations = []
+    for _ in range(2):
```
Comment (uaydonat) [inline models/demos/mnist/tests/test_perf_mnist.py:66]: It would be more stable if the test runs 100s of iterations, otherwise there might too much variation.

File: models/demos/mnist/tests/test_perf_mnist.py (hunk 7 above)
Comment (uaydonat) [reply models/demos/mnist/tests/test_perf_mnist.py:93]: the test ran in CI with 0.004, the expected inference time is 0.008, double of actual run. Even if the test slows down by 50% due to a bug, this test will not catch it. We probably need expected time to be closer to the actual run.

File: models/demos/mnist/demo/demo.py (hunk 6 above)
Comment (uaydonat) [reply models/demos/mnist/demo/demo.py:63]: The test should fail if accuracy is low

File: models/demos/mnist/tests/test_perf_mnist.py (hunk 8 above)
Comment (sabira-mcw) [reply models/demos/mnist/tests/test_perf_mnist.py]: Updated the PR with respective changes suggested and triggered CIs.

**Comment by sabira-mcw** [comment]
@tt-rkim 
Fixed the GS threshold. 
 I’ve triggered the CIs, and I’ll post the links as soon as I have them.

**Comment by sabira-mcw** [comment]
CI links 

(Single-card) Demo tests : [n150](https://github.com/tenstorrent/tt-metal/actions/runs/11968800785/job/33368423795#step:9:5420) [n300](https://github.com/tenstorrent/tt-metal/actions/runs/11968800785/job/33368423997), [n300 performance](https://github.com/tenstorrent/tt-metal/actions/runs/11968800785/job/33368424182)
All post-commit tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/11968822684) 
(Single-card) Model perf tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/11968828864) (All passed)
(Single-card) Device perf regressions : [e150](https://github.com/tenstorrent/tt-metal/actions/runs/11968816909/job/33368496864) , [n300](https://github.com/tenstorrent/tt-metal/actions/runs/11968816909/job/33368497142) 
Nightly fast dispatch tests : [e150](https://github.com/tenstorrent/tt-metal/actions/runs/11968812358/job/33368459677), [n150](https://github.com/tenstorrent/tt-metal/actions/runs/11968812358/job/33368460129), [n300](https://github.com/tenstorrent/tt-metal/actions/runs/11968812358/job/33368460725) (Mnist tests are passing)

**Comment by tt-rkim** [comment]
Approved, but do we need to skip the device perf test for grayskull?

**Comment by sabira-mcw** [comment]
@tt-rkim 

I have triggered CIs this morning, [GS_Device perf ](https://github.com/tenstorrent/tt-metal/actions/runs/12005724686/job/33463229208) is passed.

(Single-card) Demo tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12005714188/job/33463002090)
All post-commit tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12005727197)
(Single-card) Model perf tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12005729700/job/33463300093) 
(Single-card) Device perf regressions :[Link]( https://github.com/tenstorrent/tt-metal/actions/runs/12005724686)
Nightly fast dispatch tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12005717661) (Latest run for failed jobs  is in progress)

**Comment by tt-rkim** [comment]
Please ensure e2e perf test passes. Issue on main
Debug by @jbaumanTT and @pgkeller

Otherwise, keep the e2e perf test in a separate PR.

**Review by uaydonat — APPROVED** [review]

**Comment by tt-rkim** [comment]
You can alternatively try putting the test first
I will unblock, but please ensure the test passes.

**Review by tt-rkim — APPROVED** [review]

**Comment by sabira-mcw** [comment]
Model's perf is passing with today's triggered CIs. Below are the respective links .Please have a look.
(Single-card) Model perf tests:
- [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12045366458/job/33583979934#step:10:7050)(n300)(Mnist passed)
- [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12045366458/job/33583978705#step:10:9300)(e150)(Mnist passed)

**Review by tt-rkim — APPROVED** [review]

**Comment by sabira-mcw** [comment]
MNIST single-device implementation has been merged into the main .
